src/main/kotlin/com/bantolomeus/service/BookService.kt

```diff
@@ -16,84 +16,84 @@ class BookService(private val bookRepository: BookRepository,
-                }
-            }
-            books = mutableListOf(foundBook) + books.toMutableList()
+        if (foundBook.currentPage == foundBook.pagesTotal) {
+            foundBook.readTime = ((Date().time - dateFormat.parse(foundBook.dateStarted).time) / DIVISOR_FOR_DAY) + 1
+        }
+        val oldPage = foundBook.currentPage
+        foundBook.currentPage = bookUpdate.currentPage
+        val pagesRead = bookUpdate.currentPage.minus(oldPage)
+        if (pagesRead <= 0 || oldPage >= bookUpdate.currentPage || bookUpdate.currentPage > foundBook.pagesTotal) {
+            return BookUpdatesFileDTO()
         }
 
-        if (oldPage < bookUpdate.currentPage && foundBook.name == bookName
-                && bookUpdate.currentPage <= foundBook.pagesTotal) {
-            bookRepository.saveBooks(BooksFileDTO(books.toMutableList()))
-            val bookUpdates = bookUpdatesRepository.getBookUpdates()
-            val currentDate = dateFormat.format(Date())
-            val foundBookUpdate = bookUpdates.bookUpdates.filter { it.date == currentDate && it.name == bookName }
-                    .getOrElse(0) { _ -> BookUpdateOutputDTO()}
+        bookRepository.saveBooks(bookRepository.getAllBooksExcept(bookName) + foundBook)
 
-            val pagesRead = bookUpdate.currentPage.minus(oldPage)
-            if (pagesRead > 0 && foundBookUpdate.date.isEmpty()) {
-                bookUpdates.bookUpdates.add(0, BookUpdateOutputDTO(
-                        name = bookName,
-                        pagesRead = pagesRead,
-                        date = currentDate)
-                )
-                response = bookUpdatesRepository.saveBookUpdate(bookUpdates)
-                progressService.saveProgress(pagesRead)
-            } else if (foundBookUpdate.date.isNotEmpty()) {
-                val oldBookUpdates = bookUpdates.bookUpdates.filter {
-                    it.date != foundBookUpdate.date || it.name != bookName
-                }.toMutableList()
-                oldBookUpdates.add(0, BookUpdateOutputDTO(
-                        name = bookName,
-                        pagesRead = bookUpdate.currentPage.minus(oldPage).plus(foundBookUpdate.pagesRead),
-                        date = currentDate)
-                )
-                response = bookUpdatesRepository.saveBookUpdate(BookUpdatesFileDTO(oldBookUpdates))
-                progressService.saveProgress(pagesRead)
-            }
-        }
-        return response
+        // todo '.firstOrNull { it.date == currentDate && it.name == bookName }' is also something that should happen in the repo.
+        //      repository.getTodaysUpdatesForBook(foundBook)
+        // todo move exception into repo
+        //      in this case I don't see any value in running the bookmaster w/o a working BookUpdatesRepository
+        val bookUpdates = bookUpdatesRepository.getBookUpdates()
+                ?: throw IllegalStateException("BookUpdatesRepository not available but it should be")
```

Comment:
This is not true. If the value is null, then exists no file. Why not sending this instead of the current text? I could not get the needed information, that the file does not exists from this exception.

Reply:
I think it should be part of a valid application state that the repo files are accessible. They may have no content, but they should be available. In the `init` of the repo I make sure that a file gets created if there is none. So `getBookUpdates` returns `null` / this exception only is thrown when there is a problem reading the existing file which is a problem. Hence the exception. It's unintended, unexpected behavior.

Reply:
aye